</p> <p style="text-align:center;"> <img src="https://images-ext-1.discordapp.net/external/8rgDaCvLqrqclKam5YznFYmsGOTU899pxnFf406bX4I/https/i.postimg.cc/tgyfnHYs/WHOlytics.png?format=webp&quality=lossless" alt="WHOlytics" width="200" style="float: right; margin-left: 0px;"/> </p> <p style="text-align:center;"> <img src="https://i.postimg.cc/9fQzwcLM/WHO-cropped.jpg" alt="WHO" style="float: center; margin-left: 0px;"/>

## 2. 🔍 **Model Creation (Linear Regression)**  
**From exploration to model development, let's create a powerful Linear Regression model.**

---

🛠 **Key Stages in Model Creation**  
1. **🎲 Train-Test Split**: Divide the data into training and testing sets for model evaluation  
2. **➕ Feature Enrichment**: Add new features that could improve model performance  
3. **📏 Scaling**: Normalise data to ensure features contribute equally to the model  
4. **⚖️ Z-Scoring**: Standardise data to have a mean of 0 and standard deviation of 1  
5. **🔧 Feature Engineering**: Create new features or transform existing ones for better model prediction  
6. **🤖 Model Creation**: Train a Linear Regression model using the preprocessed data

---

🎯 *Linear Regression model will enable accurate predictions of life expectancy*

---

### ✅ **Let’s Build the Model!**

# Estimating the model

#### Overview
* Importing a read data
* Feature enrichment
* Train-Test Split
* Feature Engineering
* Estimating the model
* Metric analysis
* A sensitive model


#### Library Import & DataFrame Initialisation

In [31]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split

import statsmodels.api as sm
import statsmodels.tools

df = pd.read_csv('https://drive.google.com/uc?id=1nXm7P-6nDbqiFYeJdZtwN5ARSNhXI4df&export=download')

### ➕ Feature Enrichment

In [32]:
df['log_Incidents_HIV'] = np.log(df['Incidents_HIV'])        #Creates logs of variables with non-linear relationships!
df['log_GDP_per_capita'] = np.log(df['GDP_per_capita'])

**Z-normalising our variables dramatically reduced the condition number**  
Lazaridis (2007) explains that condition numbers are 'misleading' and in many cases, present spurious indications of multicolinearity [see here](https://link.springer.com/article/10.1007/s11135-005-6225-5). OLS is scale invariant and scaling does not impact multicolinearity. However, the condition number is sensitive to scaling and so scaling issues can lead to spurious inferences regarding multicolinearity from the condition number. 

In [33]:
def zscores(df, to_scale):                                                       
    def scale_column(column):                                                      
        avg = np.mean(column)  
        st = np.std(column)    
        return [(x - avg) / st for x in column]

    
    for column in to_scale:                                                       
        if column in df.columns:
            df[column + '_scaled'] = scale_column(df[column])

    return df

In [34]:
lis = ['Life_expectancy', 'Year', 'Infant_deaths', 'Adult_mortality', 'Alcohol_consumption', 'Hepatitis_B', 'Measles', 'BMI', 'Polio', 'Diphtheria',
       'Population_mln', 'Thinness_ten_nineteen_years', 'Schooling', 'Economy_status_Developed','Economy_status_Developing', 
       'Life_expectancy', 'log_Incidents_HIV','log_GDP_per_capita']

zscores(df, lis)

Country          Region  Year  Infant_deaths  Under_five_deaths  \
0       Turkiye     Middle East  2015           11.1               13.0   
1         Spain  European Union  2015            2.7                3.3   
2         India            Asia  2007           51.5               67.9   
3        Guyana   South America  2006           32.8               40.5   
4        Israel     Middle East  2012            3.4                4.3   
...         ...             ...   ...            ...                ...   
2859      Niger          Africa  2000           97.0              224.9   
2860   Mongolia            Asia  2009           23.9               28.6   
2861  Sri Lanka            Asia  2004           17.7               28.9   
2862  Lithuania  European Union  2002            7.9                9.9   
2863    Iceland  Rest of Europe  2011            2.1                2.6   

      Adult_mortality  Alcohol_consumption  Hepatitis_B  Measles   BMI  ...  \
0            105.8240                1.320           97       65  27.8  ...   
1             57.9025               10.350           97       94  26.0  ...   
2            201.0765                1.570           60       35  21.2  ...   
3            222.1965                5.680           93       74  25.3  ...   
4             57.9510                2.890           97       89  27.0  ...   
...               ...                  ...          ...      ...   ...  ...   
2859         291.8240                0.092           72       64  20.8  ...   
2860         235.2330                6.560           97       97  25.3  ...   
2861         134.8950                1.560           62       95  21.9  ...   
2862         204.0120               11.000           94       95  26.1  ...   
2863          50.5745                6.840           88       90  26.1  ...   

      BMI_scaled  Polio_scaled  Diphtheria_scaled  Population_mln_scaled  \
0       1.261475      0.696414           0.690747               0.306709   
1       0.440877      0.696414           0.690747               0.071552   
2      -1.747384     -1.293275          -1.433965               8.401854   
3       0.121756      0.364800           0.433206              -0.263267   
4       0.896765      0.497445           0.497592              -0.210798   
...          ...           ...                ...                    ...   
2859   -1.929740     -3.017672          -3.365522              -0.185736   
2860    0.121756      0.630091           0.561977              -0.249197   
2861   -1.428263      0.696414           0.690747              -0.126672   
2862    0.486466      0.696414           0.561977              -0.243554   
2863    0.486466      0.563768           0.561977              -0.266418   

      Thinness_ten_nineteen_years_scaled  Schooling_scaled  \
0                               0.007695          0.052941   
1                              -0.961328          0.652121   
2                               5.010559         -0.830060   
3                               0.187979          0.084477   
4                              -0.826115          1.629730   
...                                  ...               ...   
2859                            1.787994         -2.059956   
2860                           -0.600761          0.462906   
2861                            2.373915          0.841336   
2862                           -0.352871          1.093622   
2863                           -0.893721          1.062086   

      Economy_status_Developed_scaled  Economy_status_Developing_scaled  \
0                           -0.510454                          0.510454   
1                            1.959040                         -1.959040   
2                           -0.510454                          0.510454   
3                           -0.510454                          0.510454   
4                            1.959040                         -1.959040   
...                               ...             

## 🎲 Train-Test split

In [35]:
feature_cols = list(df.columns)                                                                 
X = df[feature_cols]                                                                            
y = df['Life_expectancy']                                                                         
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)   

## 🔧 Feature Engineering

In [36]:
def feature_eng(df):
    df = df.copy()                                                                             
    df = sm.add_constant(df)                                                                    
    df = pd.get_dummies(df, columns=['Region'], drop_first=True, prefix='Region', dtype='int') 
    return df 
X_train_fe = feature_eng(X_train)                                                              

## 🤖 Estimating the linear regression

In [37]:
#The variables used for estimation are included here 

feature_columns = ['const', 'Year_scaled', 'Infant_deaths_scaled', 'Adult_mortality_scaled','BMI_scaled', 'Polio_scaled', 'Thinness_ten_nineteen_years_scaled', 
                   'Schooling_scaled', 'log_Incidents_HIV_scaled', 'log_GDP_per_capita_scaled','Region_Asia', 'Region_Central America and Caribbean',
                   'Region_European Union', 'Region_Middle East', 'Region_North America','Region_Oceania', 'Region_Rest of Europe', 'Region_South America']

**Population and Alcohol consumption were excluded at this stage on account of being insignificant**

**Africa is excluded to avoid the dummy variable trap. Its effect on life expectancy is captured by the constant**

In [38]:
lin_reg = sm.OLS(y_train, X_train_fe[feature_columns])           
results = lin_reg.fit() 
results.summary() 

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Life_expectancy   R-squared:                       0.982
Model:                            OLS   Adj. R-squared:                  0.982
Method:                 Least Squares   F-statistic:                     7238.
Date:                Sat, 19 Apr 2025   Prob (F-statistic):               0.00
Time:                        14:48:07   Log-Likelihood:                -3809.4
No. Observations:                2291   AIC:                             7655.
Df Residuals:                    2273   BIC:                             7758.
Df Model:                          17                                         
Covariance Type:            nonrobust                                         
========================================================================================================
                                           coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
const                                   68.2364      0.080    851.109      0.000      68.079      68.394
Year_scaled                              0.1831      0.029      6.426      0.000       0.127       0.239
Infant_deaths_scaled                    -3.1212      0.079    -39.359      0.000      -3.277      -2.966
Adult_mortality_scaled                  -5.0882      0.063    -80.938      0.000      -5.211      -4.965
BMI_scaled                              -0.4840      0.049     -9.867      0.000      -0.580      -0.388
Polio_scaled                             0.1278      0.043      2.979      0.003       0.044       0.212
Thinness_ten_nineteen_years_scaled      -0.0803      0.040     -2.000      0.046      -0.159      -0.002
Schooling_scaled                         0.4661      0.060      7.724      0.000       0.348       0.584
log_Incidents_HIV_scaled                -0.2834      0.048     -5.854      0.000      -0.378      -0.188
log_GDP_per_capita_scaled                1.0065      0.058     17.505      0.000       0.894       1.119
Region_Asia                              0.4430      0.119      3.736      0.000       0.210       0.676
Region_Central America and Caribbean     1.9606      0.125     15.667      0.000       1.715       2.206
Region_European Union                    0.9536      0.145      6.560      0.000       0.669       1.239
Region_Middle East                       0.1072      0.151      0.711      0.477      -0.188       0.403
Region_North America                     1.8726      0.235      7.963      0.000       1.411       2.334
Region_Oceania                          -0.0810      0.154     -0.527      0.599      -0.383       0.221
Region_Rest of Europe                    0.6233      0.153      4.063      0.000       0.322       0.924
Region_South America                     1.6252      0.138     11.750      0.000       1.354       1.896
==============================================================================
Omnibus:                        3.548   Durbin-Watson:                   2.017
Prob(Omnibus):                  0.170   Jarque-Bera (JB):                3.166
Skew:                          -0.017   Prob(JB):                        0.205
Kurtosis:                       2.821   Cond. No.                         27.8
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 🔍 Interpreting Estimation Output

- **R² (98.2%)**:  
  98.2% of the variation in the dependent variable is explained by the variation in the independent variables.

- **Durbin-Watson Stat**:  
  The Durbin-Watson stat indicates that we **do not reject** the null hypothesis of serially independent error terms.  

- **Jarque-Bera Stat**:  
  The Jarque-Bera stat indicates that we **do not reject** the null hypothesis of residual normality.

- **Coefficient Signs**:  
  The sign of the coefficients **align with expectations**, supporting the notion of no spurious results.

- **Condition Number**:  
  The condition number is **below 30**, indicating **no multicollinearity or scaling issues**!

---


## 📊 Metric Analysis

### **Comparing the RMSE of the Test and Train Data**

- **Train RMSE**:  
  The RMSE on the training data indicates how well the model fits the training set. A lower RMSE suggests a better fit.

- **Test RMSE**:  
  The RMSE on the test data evaluates the model’s generalisability. A significant difference between the train and test RMSE could indicate overfitting or underfitting.

- **Comparison**:  
  If the RMSE of the test data is higher than that of the train data, it may indicate that the model is overfitting. Conversely, if both are similar, the model generalises well.


In [39]:
#Calculating and printing the RMSE of the test data

y_pred = results.predict(X_train_fe[feature_columns])                
rmse = statsmodels.tools.eval_measures.rmse(y_train, y_pred)         
print(rmse)

1.276115343295868


In [40]:
X_test_fe = feature_eng(X_test)                                        
X_test_fe = X_test_fe[feature_columns]                                


#Printing RMSE of test data

y_test_pred = results.predict(X_test_fe)                             
rmse = statsmodels.tools.eval_measures.rmse(y_test, y_test_pred)        
print(rmse)

1.3464955362091742


### **Our RMSE's are lower than 2**

### **What about our Variance Inflation Factors?**

In [41]:
from statsmodels.stats.outliers_influence import variance_inflation_factor      #A module to evaluate the (VIF)

cols = ['Year_scaled', 'Infant_deaths_scaled', 'Adult_mortality_scaled','BMI_scaled', 'Polio_scaled', 'Thinness_ten_nineteen_years_scaled', 'Schooling_scaled', 
        'log_Incidents_HIV_scaled', 'log_GDP_per_capita_scaled', 'Region_Asia', 'Region_Central America and Caribbean', 'Region_European Union', 'Region_Middle East', 
        'Region_North America', 'Region_Oceania', 'Region_Rest of Europe', 'Region_South America']

## Create an indexed list where we output the VIF of each of the columns.

pd.Series([variance_inflation_factor(X_train_fe[cols].values, i) for i in range(X_train_fe[cols].shape[1])], index = X_train_fe[cols].columns)

Year_scaled                             1.089977
Infant_deaths_scaled                    8.463673
Adult_mortality_scaled                  5.538780
BMI_scaled                              3.335044
Polio_scaled                            2.530260
Thinness_ten_nineteen_years_scaled      2.204706
Schooling_scaled                        4.914582
log_Incidents_HIV_scaled                3.047909
log_GDP_per_capita_scaled               4.732679
Region_Asia                             1.510099
Region_Central America and Caribbean    1.150086
Region_European Union                   1.427493
Region_Middle East                      1.274859
Region_North America                    1.055750
Region_Oceania                          1.219799
Region_Rest of Europe                   1.223996
Region_South America                    1.095288
dtype: float64

### **Multicolinearity?**

All VIF scores are below 10 which is widely cited as the threshold beyond which multicolinearity is deemed problematic. For example, see [here](https://link.springer.com/book/10.1007/978-1-0716-1418-1) or [here.](https://www.drnishikantjha.com/papersCollection/Multivariate%20Data%20Analysis.pdf)


## A Sensitive Model

**Some countries have expressed concern over sharing sensitive data.**  
To address these concerns, those countries can use this model.

### **Excluded Variables**  
The following variables have been excluded from the model to protect sensitive information:

- **Infant Deaths**  
- **Thinness in 10-19 Year Olds**  
- **HIV**  
- **Schooling**

In [42]:
feature_columns_sensitive = ['const', 'Year_scaled', 'Adult_mortality_scaled','BMI_scaled', 'Polio_scaled', 
                    'log_GDP_per_capita_scaled','Region_Asia', 'Region_Central America and Caribbean',
                   'Region_European Union', 'Region_Middle East', 'Region_North America','Region_Oceania', 'Region_Rest of Europe', 'Region_South America']

In [43]:
lin_reg_sensitive = sm.OLS(y_train, X_train_fe[feature_columns_sensitive])           
results_sensitive = lin_reg_sensitive.fit()                                          

In [44]:
y_pred_sensitive = results_sensitive.predict(X_train_fe[feature_columns_sensitive])               
rmse_sensitive = statsmodels.tools.eval_measures.rmse(y_train, y_pred_sensitive)    
print(rmse_sensitive)

1.7279157339297657


In [45]:

X_test_fe = feature_eng(X_test)
X_test_fe = X_test_fe[feature_columns_sensitive]



y_test_pred = results_sensitive.predict(X_test_fe)
rmse = statsmodels.tools.eval_measures.rmse(y_test, y_test_pred)
print(rmse)

1.9110964279138003


### Interpreting estimation output
* The R-squared has fallen by 1.3%
* The Durbin-Watson stat indicates that we do not reject the null hypothesis of serially independent error terms.
* The Jarque-Bera stat indicates that the residuals are now non-normal, so we cant trust the confindence intervals drawn from the t-distribution because this assumes residual normality. However, this does not affect forecasting as the coefficient estimates are not biased.
* The sign of the coefficients align with expectations supporting the notion of no spurious results.
* The condition number is below 30, indicating no multicolinearity or scaling issues!
* The RMSE's are higher, indicating that the model is estimating the dependent variable with less power.
* Overall, it is still a good model. 

### 🔍 Interpreting Estimation Output

- **R² (96.7%)**:  
  The R-squared has fallen by **1.3%**, indicating a slight reduction in the model's explanatory power.

- **Durbin-Watson Stat**:  
  The Durbin-Watson stat indicates that we **do not reject** the null hypothesis of serially independent error terms.

- **Jarque-Bera Stat**:  
  The Jarque-Bera stat indicates that the residuals are now **non-normal**, meaning we can’t trust the confidence intervals drawn from the t-distribution (which assumes residual normality). However, this does not affect forecasting, as the coefficient estimates remain unbiased.

- **Coefficient Signs**:  
  The sign of the coefficients **align with expectations**, supporting the notion of no spurious results.

- **Condition Number**:  
  The condition number is still **below 30**, indicating **no multicollinearity or scaling issues**.

- **RMSE**:  
  The RMSE values are **higher**, indicating that the model is estimating the dependent variable with **less power**.

- **Overall**:  
  Despite these changes, it is still a **good, accurate model**.
